### ISSUE TITLE
my flynn cluster is frozen?

### ISSUE THREAD

**Issue by christiangenco** [flynn/flynn#2362]
After trying to add a new domain name yesterday (with `flynn route add http -s dbinbox-web -c ../dbinbox.com.csr -k ../dbinbox.com.key dbinbox.com`) my cluster started acting strangely:

* my deployed app now shows no domains (not even the `dbinbox.6w5k.flynnhub.com` domain where it's currently deployed) <img height="100" alt="screen shot 2016-01-10 at 4 10 27 pm" src="https://cloud.githubusercontent.com/assets/179361/12224392/b995044a-b7b4-11e5-8a60-53de951306c6.png">
* deployments (via `git push flynn`) take longer than they used to, and [time out before completing](https://gist.github.com/christiangenco/1392e29165dca01f42d2)
* any changes made on the dashboard just say "please wait" and never complete:

<img width="790" alt="screen shot 2016-01-10 at 4 11 50 pm" src="https://cloud.githubusercontent.com/assets/179361/12224402/ebd0dcf4-b7b4-11e5-9236-7af2c13bc32f.png">

I'd imagine something just needs to be restarted, but I'm not sure what. I hesitate to restart the hosts, because [the deployed app is still live](https://dbinbox.6w5k.flynnhub.com/) and if this happens again in production I'd like to troubleshoot without taking down all apps on the cluster.

Here's the output of [`flynn-host collect-debug-info`](https://gist.github.com/anonymous/cbdd806c9e8ddea18bcd) and [`flynn-host fix`](https://gist.github.com/christiangenco/23044f9fa177a884d990).

I'm not sure what to be looking for in the debug info, but I noticed that [free memory looks really low](https://gist.github.com/anonymous/cbdd806c9e8ddea18bcd#file-0-debug-output-log-L671). I'm playing with flynn on one 2GB box and two 1GB boxes right now (the minimum recommended ram), but plan to have them all on 2GB in production.

**Comment by josephglanville** [comment]
It looks like the network your hosts are running on is either very slow on the machines the hosts are running on are slow enough that it causes the `discoverd` service to have trouble communicating.

Could you provide any additional information about what kind of environment your machines are hosted in?

**Comment by christiangenco** [comment]
@josephglanville Interesting. Is that the [`2016/01/10 05:19:37 discoverd: heartbeat blobstore (100.100.9.7:80) failed: unknown_error: Something went wrong`](https://gist.github.com/anonymous/cbdd806c9e8ddea18bcd#file-blobstore-web-flynn01-fe943233-4a0f-46cf-a4b0-258d6453f465-log-L55)?

All three are hosted on https://chunkhost.com/ in different zones within the the Peer One data center in downtown Los Angeles:

<img width="619" alt="screen shot 2016-01-10 at 7 53 10 pm" src="https://cloud.githubusercontent.com/assets/179361/12225605/ed8845ae-b7d3-11e5-971e-f4bbe8ebc7bb.png">

Here's what they say about CPU:

> ### How much CPU do I get?
> Xen (the virtualization software that runs your chunk) is very good about CPU scheduling. 99% of the time CPU isn’t really an issue, but if somehow all the chunks on a machine were all trying to max out the processor at the same time, you would be guaranteed the same ratio of CPU as you have of the total machine memory. Again, remember you can always burst to the unused portion of all twelve cores!

And the network:

> ### How fast is the uplink for my chunk?
> Depending on how much other chunks are using, the chunk could get something approaching 1gbps per second down. Upload is 100mbps.
> You can test it yourself with our download test page: http://speedtest-lax1.chunkhost.com/100mb.test

I'm with them because their pricing is fantastic, but can try switching to linode if you think their hardware is the issue.

<img width="377" alt="screen shot 2016-01-10 at 7 57 27 pm" src="https://cloud.githubusercontent.com/assets/179361/12225640/746e8a1a-b7d4-11e5-8d77-5c27779fceee.png">

**Comment by titanous** [comment]
What is the latency between the two datacenters? The raft consensus cluster is pretty latency-sensitive so typically you'd want to run all servers in the the same DC.

**Comment by christiangenco** [comment]
Ahh interesting - there's a significant difference in ping latency between computers not in the same datacenter.

Here's what I got pinging each machine from each machine (times in miliseconds):

<table>
  <tr>
    <td></td>
    <td>flynn1</td>
    <td>flynn2</td>
    <td>flynn3</td>
  </tr>
  <tr>
    <td>flynn1</td>
    <td>0.067</td>
    <td><b>26.413</b></td>
    <td><b>26.380</b></td>
  </tr>
  <tr>
    <td>flynn2</td>
    <td><b>26.366</b></td>
    <td>0.065</td>
    <td>0.599</td>
  </tr>
    <tr>
    <td>flynn3</td>
    <td><b>26.471</b></td>
    <td>0.638</td>
    <td>0.057</td>
  </tr>
</table>

ie: latency between data centers is over 40x slower than latency in the same datacenter.

I suppose the question now is: is there an elegant way to remove `flynn01` from the cluster and add a new host, or do I need to install everything from scratch?

**Comment by titanous** [comment]
Yeah, that kind of latency will break things as you've seen. The easiest thing to do is to install a new cluster, moving hosts around is pretty rough right now.

**Comment by christiangenco** [comment]
:+1: Will do! Thank you immensely for your help - I never would've suspected network latency as the culprit.

**Comment by christiangenco** [comment]
Moving all nodes to the same zone seems to have fixed everything. Thanks @titanous again for your help!

**Comment by christiangenco** [comment]
Well shoot, I spoke too soon. After trying to add my SSL certificates again, the cluster froze again. It's exactly the same situation as in [my first issue](https://github.com/flynn/flynn/issues/2362#issue-125849609) (all commands on the client and dashboard are just hanging).

Here's the [`flynn-host collect-debug-info`](https://gist.github.com/e8752fd9fbf0833cf265).

As a bonus of moving things to the same zone, the things that aren't frozen are going much faster :smile:

**Comment by titanous** [comment]
Looks like the router is breaking due to a bad certificate:

```text
2016/01/11 11:29:12 router: sync error: crypto/tls: failed to parse certificate PEM data
2016/01/11 11:29:14 router: sync error: crypto/tls: failed to parse certificate PEM data
2016/01/11 11:29:17 router: sync error: crypto/tls: failed to parse certificate PEM data
2016/01/11 11:29:19 router: sync error: crypto/tls: failed to parse certificate PEM data
```

From your original command, you are providing a `.csr` file which is a certificate *request*, not a certificate. I suspect that this is the problem. We need to add verification of the TLS certs/keys before persisting it to the database.

**Comment by christiangenco** [comment]
@titanous Ahh! Now that makes a lot of sense.

Is there a clever way to connect to the flynn database and remove the certificates without reinstalling my entire cluster? Here's a record of everything I've tried:

I found the postgres processes from `flynn-host ps | grep postgres`:

```
flynn03-f17a263e-d864-43c3-bfc2-6267ef7876bc  running  2 days ago  postgres        web
flynn04-dbd82c9b-70b9-4f3c-8cf5-06c31a5abae2  running  2 days ago  postgres        web
flynn02-2aa8ae94-64f9-4b89-86be-a4d7a26e4faa  running  2 days ago  postgres        postgres
flynn04-b3dc85d3-7e01-4882-8f44-ba234e1a97c6  running  2 days ago  postgres        postgres
flynn03-c381a5a5-1620-49e8-8282-d89385d8f28d  running  2 days ago  postgres        postgres
```

The web process [looks like it's just an internal API for creating the database](https://github.com/flynn/flynn/blob/48194db91a768bff45095a27ff81dec8970f3287/appliance/postgresql/api/server.go), but the postgres processes look like replicated postgres instances I can connect to:

```
$ flynn-host inspect flynn02-2aa8ae94-64f9-4b89-86be-a4d7a26e4faa
ID                         flynn02-2aa8ae94-64f9-4b89-86be-a4d7a26e4faa
Entrypoint
Cmd                        postgres
Status                     running
StartedAt                  2016-01-11 11:01:38.940399579 +0000 UTC
EndedAt
ExitStatus                 <nil>
IP Address                 100.100.13.2
flynn-controller.app       e08277a8-4484-4e93-8c51-1e85a70acf33
flynn-controller.app_name  postgres
flynn-controller.release   11a5d07a-62fe-452a-90d0-c537db881c00
flynn-controller.type      postgres
flynn-system-app           true
ENV[PGPASSWORD]            af0cb9545f83c7ad3c94220d2ce25d3b
ENV[PORT_0]                5432
ENV[FLYNN_APP_ID]          e08277a8-4484-4e93-8c51-1e85a70acf33
ENV[FLYNN_JOB_ID]          flynn02-2aa8ae94-64f9-4b89-86be-a4d7a26e4faa
ENV[PGDATABASE]            postgres
ENV[PGUSER]                flynn
ENV[EXTERNAL_IP]           100.100.13.2
ENV[FLYNN_PROCESS_TYPE]    postgres
ENV[FLYNN_RELEASE_ID]      11a5d07a-62fe-452a-90d0-c537db881c00
ENV[PGHOST]                leader.postgres.discoverd
ENV[PORT]                  5432
ENV[SINGLETON]             false
ENV[FLYNN_APP_NAME]        postgres
ENV[FLYNN_POSTGRES]        postgres
```

I was able to connect to that with: `psql -h 100.100.13.2 -p 5432 -d postgres -U flynn --password` and the password `af0cb9545f83c7ad3c94220d2ce25d3b`.

Then I listed the databases:

```                      
postgres=# \z                                  
List of databases
               Name               |              Owner               | Encoding |   Collate   |    Ctype    |   Access privileges
----------------------------------+----------------------------------+----------+-------------+-------------+-----------------------
 2149ba3fe66a5ca24e4ebc7830f1dd77 | f45111b0d93069f9e4d961a3f8ce56bf | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 e4f660acffc69b10a645563d2ef34303 | 9b32fbd0f415bfdfacbe6384a3edc00b | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 f2bc3d5cd142a6ff8c90aeff2a49b246 | 5ebe63bfdf9a10309935ec47469a347c | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 postgres                         | postgres                         | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 template0                        | postgres                         | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
                                  |                                  |          |             |             | postgres=CTc/postgres
 template1                        | postgres                         | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
                                  |                                  |          |             |             | postgres=CTc/postgres
(6 rows)
```

and tried poking around in that first one:

```
postgres=# \connect 2149ba3fe66a5ca24e4ebc7830f1dd77
2149ba3fe66a5ca24e4ebc7830f1dd77=# \z                                 Access privileges
 Schema |       Name        | Type  | Access privileges | Column access privileges
--------+-------------------+-------+-------------------+--------------------------
 public | http_routes       | table |                   |
 public | schema_migrations | table |                   |
 public | tcp_routes        | table |                   |
(3 rows)
```

Bingo! `http_routes`! Let's see what's in there…

```
2149ba3fe66a5ca24e4ebc7830f1dd77=# select id, domain from http_routes;
                  id                  |               domain
--------------------------------------+------------------------------------
 41e32830-5d9e-44d9-97ae-5d806ae13803 | git.flynn.gen.co
 76427838-5c09-434a-b9f7-0076fb702630 | controller.flynn.gen.co
 9fa27c50-eec1-4cc6-b60f-fc2d58c183a2 | dashboard.flynn.gen.co
 7442da76-adb2-4c48-b310-7ffa94e91cce | status.flynn.gen.co
 44806653-7375-4b25-864f-ff81684937d4 | dbinbox.flynn.gen.co
 97d27132-9786-4ae6-b4c6-a5cd4b014f39 | jellyfish-drag-guelph.flynn.gen.co
 0ed3fb32-da7c-4d14-bc5b-861616267257 | dbinbox.com
 8e5083c4-38bb-4acb-9803-cdd06c8e7926 | dbinbox.com
(8 rows)
```

Bingo. And the second `dbinbox.com` was the one that had a value at `tls_cert` and `tls_key`, so I tried to delete it:

```
2149ba3fe66a5ca24e4ebc7830f1dd77=# delete from http_routes where id='8e5083c4-38bb-4acb-9803-cdd06c8e7926';
ERROR:  cannot execute DELETE in a read-only transaction
```

But apparently this node is a [read replica](http://serverfault.com/questions/630753/how-to-change-postgresql-database-from-read-only-to-writable), so I tried the other two postgres processes and the last one was master. Even on that one, and even with a transaction, I couldn't delete the record:

```
2149ba3fe66a5ca24e4ebc7830f1dd77=# BEGIN; SET TRANSACTION READ WRITE;
2149ba3fe66a5ca24e4ebc7830f1dd77=#   delete from http_routes where id='8e5083c4-38bb-4acb-9803-cdd06c8e7926';
2149ba3fe66a5ca24e4ebc7830f1dd77=# COMMIT;
ERROR:  record "new" is not assigned yet
DETAIL:  The tuple structure of a not-yet-assigned record is indeterminate.
CONTEXT:  PL/pgSQL function check_http_route_update() line 8 at IF
```

So I just deleted its `tls_cert` and `tls_key` values:

```
UPDATE http_routes SET tls_cert=NULL, tls_key=NULL WHERE id='8e5083c4-38bb-4acb-9803-cdd06c8e7926';
```

That looks like it cleared out the SSL stuff, but the dashboard is still frozen (and things like `flynn route` from my dev machine still timeout). The router is probably caching it or something? Let's get the router's process IDs again:

```
root@flynn04:~# flynn-host ps | grep router
flynn03-3ff9230c-405c-488e-805c-3b445429902e  running  2 days ago      router          app
flynn02-d7f0cff4-80e3-4c8f-9efe-955833998c55  running  2 days ago      router          app
flynn04-3e8282a5-0047-4b96-a9dd-bde7ed443d0f  running  2 days ago      router          app
```

Nothing interesting in the logs:

    root@flynn04:~# flynn-host log flynn04-3e8282a5-0047-4b96-a9dd-bde7ed443d0f | less
    # nothing interesting

Let's try to kill them?

```
root@flynn04:~# flynn-host stop flynn03-3ff9230c-405c-488e-805c-3b445429902e
flynn03-3ff9230c-405c-488e-805c-3b445429902e stopped
```

Looks like that killed it:

```
root@flynn04:~# flynn-host ps | grep router
flynn02-d7f0cff4-80e3-4c8f-9efe-955833998c55  running  2 days ago      router          app
flynn04-3e8282a5-0047-4b96-a9dd-bde7ed443d0f  running  2 days ago      router          app
```

Anything new? On my development machine:

```
$ flynn route
Get https://controller.flynn.gen.co/apps/dbinbox/routes: dial tcp 76.74.170.138:443: connection refused
```

Oh hey that's new! Let's try to bring it back up?

```
root@flynn04:~# flynn-host fix --min-hosts 3
INFO[01-14|05:43:41] found expected hosts                     n=3
INFO[01-14|05:43:41] ensuring discoverd is running on all hosts
INFO[01-14|05:43:41] checking flannel
INFO[01-14|05:43:41] flannel looks good
INFO[01-14|05:43:41] waiting for discoverd to be available
INFO[01-14|05:43:41] checking for running controller API
INFO[01-14|05:43:41] found running controller API instances   n=2
INFO[01-14|05:43:41] found controller instance, checking critical formations
INFO[01-14|05:43:41] checking postgres
INFO[01-14|05:43:41] found running postgres leader
INFO[01-14|05:43:41] found 3 running postgres instances
INFO[01-14|05:43:41] getting postgres status
INFO[01-14|05:43:41] postgres claims to be read-write
INFO[01-14|05:43:41] checking for running controller API
INFO[01-14|05:43:41] found running controller API instances   n=2
INFO[01-14|05:43:41] found controller instance, checking critical formations
INFO[01-14|05:43:41] scheduler looks up, moving on
INFO[01-14|05:43:41] cluster fix complete
```

Huh - I don't think it's checking the router. But [it looks like it should be](https://github.com/flynn/flynn/blob/df35fe2ab64e0c839140af7af74da3c89c37d346/host/cli/fix.go#L279). Let's try killing the other two `router` processes and try again:

```
root@flynn04:~# flynn-host stop flynn02-d7f0cff4-80e3-4c8f-9efe-955833998c55
flynn02-d7f0cff4-80e3-4c8f-9efe-955833998c55 stopped
```

Now `flynn route` is hanging again. Let's stop the other one!

```
root@flynn04:~# flynn-host stop flynn04-3e8282a5-0047-4b96-a9dd-bde7ed443d0f
flynn04-3e8282a5-0047-4b96-a9dd-bde7ed443d0f stopped
```

Now `flynn route` is refusing the connection again:

    Get https://controller.flynn.gen.co/apps/dbinbox/routes: dial tcp 69.90.132.83:443: connection refused

Let's do another fix!

```
root@flynn04:~# flynn-host fix --min-hosts 3
INFO[01-14|05:48:45] found expected hosts                     n=3
INFO[01-14|05:48:45] ensuring discoverd is running on all hosts
INFO[01-14|05:48:45] checking flannel
INFO[01-14|05:48:45] flannel looks good
INFO[01-14|05:48:45] waiting for discoverd to be available
INFO[01-14|05:48:45] checking for running controller API
INFO[01-14|05:48:45] found running controller API instances   n=2
INFO[01-14|05:48:45] found controller instance, checking critical formations
INFO[01-14|05:48:45] checking postgres
INFO[01-14|05:48:45] found running postgres leader
INFO[01-14|05:48:45] found 3 running postgres instances
INFO[01-14|05:48:45] getting postgres status
INFO[01-14|05:48:45] postgres claims to be read-write
INFO[01-14|05:48:45] checking for running controller API
INFO[01-14|05:48:45] found running controller API instances   n=2
INFO[01-14|05:48:45] found controller instance, checking critical formations
INFO[01-14|05:48:45] scheduler looks up, moving on
INFO[01-14|05:48:45] cluster fix complete
```

...huh, I thought that would've done it. Now everything's very down - the web dashboard was loading up until that last `flynn-host stop`.

---

tl;dr: I tried to delete the bad certificate but nothing changed, then I manually stopped all the router instances to try to restart them but couldn't bring them back up (not even with `flynn-host fix`).

I reinstalled flynn on my server cluster, crossed my fingers, and re-added the domain with `flynn route add http -s dbinbox-web -c ../dbinbox.com.bundle.crt -k ../dbinbox.com.key dbinbox.com` and... nothing crashed! The web UI even live updated right after I ran the command - React is neat.

Then, to test the certificate, I edited my `/etc/hosts` file on my development laptop to spoof changing the DNS to flynn:

```
$ ping -c 1 dbinbox.flynn.gen.co                
PING flynn.gen.co (76.74.170.138): 56 data bytes
64 bytes from 76.74.170.138: icmp_seq=0 ttl=53 time=71.772 ms

--- flynn.gen.co ping statistics ---
1 packets transmitted, 1 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 71.772/71.772/71.772/0.000 ms
76.74.170.183 dbinbox.com
$ echo "76.74.170.183 dbinbox.com" | sudo tee -a /etc/hosts
```

And that totally worked! So I'm good to go for real this time, but I don't yet have complete faith flynn could survive one of my servers rebooting (or that I know enough about flynn to bring it back up), so I'll be doing a little more testing before switching the production switch.
